common/web/gesture-recognizer/src/engine/headless/gestures/matchers/gestureMatcher.ts

```diff
@@ -0,0 +1,293 @@
+import { ComplexGestureSource } from "../../complexGestureSource.js";
+import { SimpleGestureSource, SimpleGestureSourceSubview } from "../../simpleGestureSource.js";
+
+import { GestureModel, GestureResolution, GestureResolutionSpec, RejectionDefault, ResolutionItemSpec } from "../specs/gestureModel.js";
+
+import { ManagedPromise, TimeoutPromise } from "@keymanapp/web-utils";
+import { FulfillmentCause, PathMatcher } from "./pathMatcher.js";
+
+//
+// TODO:  Next up - let's unit-test this bad boy!
+//
+
+export interface MatchResult<Type> {
+  matched: boolean,
+  action: GestureResolution<Type>
+}
+
+export interface MatchResultSpec {
+  matched: boolean,
+  action: GestureResolutionSpec
+}
+
+export class GestureMatcher<Type> {
+  private sustainTimerPromise?: TimeoutPromise;
+  public readonly model: GestureModel<Type>;
+
+  public readonly pathMatchers: PathMatcher<Type>[];
+  private predecessor?: GestureMatcher<Type>;
+
+  private readonly publishedPromise: ManagedPromise<MatchResult<Type>>; // unsure on the actual typing at the moment.
+  private _result: MatchResult<Type>;
+
+  private baseSource: ComplexGestureSource<Type>;
+
+  public get promise() {
+    return this.publishedPromise.corePromise;
+  }
+
+  constructor(model: GestureModel<Type>, sourceObj: ComplexGestureSource<Type> | GestureMatcher<Type>) {
+    /* c8 ignore next 5 */
+    if(!model || !sourceObj) {
+      throw new Error("Construction of GestureMatcher requires a gesture-model spec and a source for related contact points.");
+    } else if(!model.sustainTimer && sourceObj instanceof ComplexGestureSource && sourceObj.touchpoints.length == 0) {
+      throw new Error("If the provided gesture-model spec lacks a sustain timer, there must be an active contact point.");
+    }
+
+    // We condition on ComplexGestureSource since some unit tests mock the other type without
+    // instantiating the actual type.
+    const predecessor = sourceObj instanceof ComplexGestureSource<Type> ? null : sourceObj;
+    const source = predecessor ? null : (sourceObj as ComplexGestureSource<Type>);
+
+    this.baseSource = predecessor?.baseSource || source;
+
+    this.predecessor = predecessor;
+    this.publishedPromise = new ManagedPromise();
+
+    this.model = model;
+    if(model.sustainTimer) {
+      this.sustainTimerPromise = new TimeoutPromise(model.sustainTimer.duration);
+      this.sustainTimerPromise.then((elapsed) => {
+        const shouldResolve = model.sustainTimer.expectedResult == elapsed;
+        this.finalize(shouldResolve, 'timer');
+      });
+    }
+
+    this.pathMatchers = [];
+
+    let sourceTouchpoints: SimpleGestureSource<Type>[];
+    if(source) {
+      sourceTouchpoints = source.touchpoints;
+    } else {
+      sourceTouchpoints = predecessor.pathMatchers.map((matcher) => matcher.source);
+    }
```

Comment:
```suggestion
    const sourceTouchpoints = source
      ? source.touchpoints
      : predecessor.pathMatchers.map(matcher => matcher.source);
```

Allows us to `const` the var but not essential. Probably can eliminate the type as well then?

Reply:
I'd like to keep the type, as it provides a clear compile-time assertion that can otherwise cause weird errors down the line.  I've had a few spots before where TS happily provided a union-type I didn't expect that caused side-effect errors later.

The part about turning it const, though - can do.